barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -509,4 +514,95 @@ namespace avm_alu(256);
     // 128-bit multiplication and CAST need two rows in ALU trace. We need to ensure
     // that another ALU operation does not start in the second row.
     #[TWO_LINE_OP_NO_OVERLAP]
-    (op_mul * ff_tag + op_cast) * alu_sel' = 0;
\ No newline at end of file
+    (op_mul * ff_tag + op_cast) * alu_sel' = 0;
+
+    // ========= SHIFT LEFT/RIGHT OPERATIONS ===============================
+    // Given inputs to a shift operation, a & b, and a memory tag, mem_tag.
+    // Split a into Big Endian hi and lo limbs, a_hi and a_lo, and the number of bits represented by the memory tag, t.
+    // QUESTION: SHOULD B BE CONSTRAINED TO BE U8 -> i.e. when would we shift more than 255 bits when the max number of bits of a is 128bits?
+    // If we are shifting by more than the bit length represented by the memory tag, the result is trivially zero
+    //
+    // === Steps when performing SHR
+    // (1) Prove the correct decomposition: a_hi * 2**b + a_lo = a
+    // (2) Range check a_hi < 2**(t-b) && a_lo < 2**b, ensure we have not overflowed the limbs during decomp
+    // (3) Return a_hi
+    //
+    //  <--(t-b) bits --> |   <-- b bits -->
+    // -------------------|-------------------
+    // |      a_hi        |       a_lo       | --> a
+    // ---------------------------------------
+    //
+    // === Steps when performing SHL
+    // (1) Prove the correct decomposition: a_hi * 2**(t-b) + a_lo = a
+    // (2) Range check a_hi < 2**b && a_lo < 2**(t-b)
+    // (3) Return a_lo * 2**b
+    // 
+    //  <-- b bits -->   | <-- (t-b) bits -->
+    // ------------------|-------------------
+    // |      a_hi       |      a_lo        | --> a
+    // --------------------------------------
+
+    // TODO: Possibly optimised with variable length checks
+    // Check that a_lo and a_hi are range checked such that that:
+    // SHR: a_hi < 2**(t - b) && a_lo < 2**b
+    // SHL: a_hi < 2**b && a_lo < 2**(t - b)
+
+    // In lieu of a variable length check, we can utilise 2 fixed range checks instead.
+    // Given the dynamic range check of 0 <= a_hi <= 2**(t-b), where b < t
+    // (1) 0 <= a_hi <= 2**t
+    // (2) 0 <= 2**(t - b)  - a_hi <= 2**t
+    // Note that (1) is guaranteed elsewhere through the tagged memory model, so we focus on (2) here. 
```

Comment:
Where is (2) enforced? Same for a_lo < 2**b
By the range check inherited from LT/LTE, we range check a_lo and a_hi to be 128 bits.